lib/rucio/transfertool/fts3_plugins.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Callable, Optional
+import json
+import sys
+
+from rucio.core.plugins import PolicyPackageAlgorithms
+from rucio.common.config import config_get_int, config_get_items
+from configparser import NoSectionError
+
+from rucio.common.exception import InvalidRequest
+
+
+class FTS3MetadataPlugin(PolicyPackageAlgorithms):
+    """
+    Add a "archive_metadata" field to a file's transfer parameters.
+    Plugins are registered during initialization and called during a transfer with FTS3
+    """
+
+    ALGORITHM_NAME = "fts3_plugins"
+    HINTS_NAME = "fts3_plugins_init"
+    DEFAULT = "def"
+
+    def __init__(self, policy_algorithm: str) -> None:
+        """
+        :param policy_algorithm: policy algorithm indentifier - choose from any of the policy package algorithms registered under the `fts3_plugins` group.
+        :type policy_algorithm: str
+        """
+        super().__init__()
+        self.register("activity", func=self._activity_hints, init_func=self._init_activity_hints)
+        self.register(self.DEFAULT, func=lambda x: self._collocation(self._default, x))
+        self.register("cms_collocation", func=lambda x: self._collocation(self._cms_collocation, x))
+        self.register("test", func=lambda x: self._collocation(self._test_collocation, x))
+
+        self.transfer_limit = config_get_int(
+            "transfers",
+            option="metadata_byte_limit",
+            raise_exception=False,
+            default=4096,
+        )
+
+        # Use a default if the algorithm isn't supported
+        if not self._supports(self.ALGORITHM_NAME, policy_algorithm):
+            logging.debug(f"Policy Algorithm {policy_algorithm} not found, ignoring.")
+            policy_algorithm = self.DEFAULT
+
+        # If the policy has a supplied and registered init function
+        if self._supports(self.HINTS_NAME, policy_algorithm):
+            self._get_one_algorithm(self.HINTS_NAME, name=policy_algorithm)()
+
+        self.set_in_hints = self._get_one_algorithm(self.ALGORITHM_NAME, name=policy_algorithm)
+
+    @classmethod
+    def register(cls, name: str, func: Callable, init_func: Optional[Callable] = None) -> None:
+        """
+        Register a fts3 transfer plugin
+
+        :param name: name to register under
+        :type name: str
```

Comment:
I think the `type` notes in the docstring might be unnecessary given that the types are already mentioned in the function arguments